## Homework 1
Probability Review and Priors

Name: Hyun Suk Ryoo (Max Ryoo)

Computing ID: hr2ee

#### Question 1

(15) You are a data scientist and are choosing between three approaches, A, B,
and C, to a problem. With approach A you will spend a total of 4 days coding
and running an algorithm and it will not produce useful results. With approach
B you will spend a total of 3 days coding and running an algorithm and it will
not produce useful results. With approach C you will spend 1 day coding and
running an algorithm and it will produce useful results. You are equally likely to
choose among unselected options. What is the expected time in days for you to
obtain the results you are looking for, if you continue to select an unselected
option when you do not obtain useful results? What is the variance on this time?

#### Solution 1

For this problem you can think of the the possible choices as shown below.

Starting with A
- $A \rightarrow B \rightarrow C$
- $A \rightarrow C$

Starting with B
- $B \rightarrow A \rightarrow C$
- $B \rightarrow C$

Starting with C
- $C$

We always end at $C$ given the information that the algorithm will produce useful results. We are also given that the approaches $A$, $B$, and $C$ are all equally likely. First lets compute how many days each option (steps) will take

Starting with A
- $A \rightarrow B \rightarrow C = 4 \rightarrow 3 \rightarrow 1 = 8$ days
- $A \rightarrow C = 4 \rightarrow 1 = 5$ days

Starting with B
- $B \rightarrow A \rightarrow C = 3 \rightarrow 4 \rightarrow 1 = 8$ days
- $B \rightarrow C = 3 \rightarrow 1 = 4$ days

Starting with C
- $C \rightarrow 1$ days

Additionally we need to compute the probabilities that each approach will be. We are alos given the informatino taht we will choose among the unselected options. 

Starting with A
- $A \rightarrow B \rightarrow C = \frac{1}{3} \rightarrow \frac{1}{2} \rightarrow 1 = \frac{1}{6}$
- $A \rightarrow C = \frac{1}{3} \rightarrow \frac{1}{2} = \frac{1}{6}$

Starting with B
- $B \rightarrow A \rightarrow C = \frac{1}{3} \rightarrow \frac{1}{2} \rightarrow 1 = \frac{1}{6}$
- $B \rightarrow C =  \frac{1}{3} \rightarrow  \frac{1}{2} =  \frac{1}{6}$

Starting with C
- $C \rightarrow  \frac{1}{3}$

Given all the days and propbabilities we can find the expected number of days with the given probabilites.

$D(A_1)$ = The days for the first option starting with A

$P(A_1)$ = The probability for the first option starting with A

$E(X) = D(A_1) \cdot P(A_1) + D(A_2) \cdot P(A_2) + D(B_1) \cdot P(B_1) + D(B_2) \cdot P(B_2) + D(C_1) \cdot P(C_1)$

$E(X) = 8 \cdot \frac{1}{6} + 5 \cdot \frac{1}{6} + 8 \cdot \frac{1}{6} + 4 \cdot \frac{1}{6} + 1 \cdot \frac{1}{3}$

$E(X) = \frac{8}{6} + \frac{5}{6} + \frac{8}{6} + \frac{4}{6} + \frac{2}{6}$

$E(X) = \frac{27}{6} = 4.5$

In order to find the variance let us map the data into data point values.

Expected Days = 4.5

Paths = [ABC, AC, BAC, BC, C]

Proabilities = [$\frac{1}{6}$, $\frac{1}{6}$, $\frac{1}{6}$, $\frac{1}{6}$,$\frac{1}{3}$]

Actual Days = [8, 5, 8, 4, 1]

We can compute the variance of X to be the following formulation.

$VAR[X] = E[(X-E[X])^2] = \Sigma_{x\in dom(X)} (x-E[x])^2 p(x)$

In [12]:
probs = [1/6, 1/6, 1/6, 1/6, 1/3]
days = [8,5,8,4,1]
expected = 0
for i in range(len(days)):
    expected = probs[i]*days[i] + expected
expected

4.5

In [17]:
probs = [1/6, 1/6, 1/6, 1/6, 1/3]
actual = 8,5,8,4,1
var = 0
for i in range(len(actual)):
    var = var + (actual[i] - expected)**2 * probs[i]
var

8.25

From the computation above and code computation we can see that for this problem the expected time to produce results is **4.5** with a variance of **8.25**

#### Question 2
(15) Suppose if it is sunny or not in Charlottesville depends on the weather of
the last three days. Show how this can be modeled as a Markov chain.

#### Solution 2

The decision for tomorrow's weather condition depends on the last three days. Let us say for the sake of explanation we have the following weather data where S= Sunny and N = Not Sunny.

In [58]:
import random
sampleSize = 600
weather = []
while sampleSize:
    random.seed(sampleSize)
    n = random.randint(1,100)
    if n > 50:
        weather.append("S")
    else:
        weather.append("N")
    sampleSize = sampleSize -1
weather[0:10]

['N', 'S', 'N', 'N', 'S', 'N', 'N', 'N', 'N', 'S']

Since the weather condition is dependent on the last three days the following options are possible for the three day sequence.

- Sunny, Sunny, Sunny
- Sunny, Sunny, Not Sunny
- Sunny, Not Sunny, Sunny
- Sunny, Not Sunny, Not Sunny
- Not Sunny, Sunny, Sunny
- Not Sunny, Sunny, Not Sunny
- Not Sunny, Not Sunny, Sunny
- Not Sunny, Not Sunny, Not Sunny

For each of these possibilites it is possible the next day will be sunny or not sunny.

We can consider making a dictionary value where the KEY is the preceding two days and the value is a tuple where the tuple is (number of occurences, the next day is sunny).

In [61]:
count_dict = {
    "SSS": (0,0),
    "SSN" : (0,0),
    "SNS" : (0,0),
    "SNN" : (0,0),
    "NSS" : (0,0),
    "NSN" : (0,0),
    "NNS" : (0,0),
    "NNN" : (0,0)
}
for i in range(3,len(weather)):
    previous = "".join(weather[i-3:i])
    if previous not in count_dict.keys():
        print(previous)
        print(weather[i])
        break
    success, total = count_dict[previous]
    if weather[i] == "S":
        count_dict[previous] = (success+1,total+1)
    else:
        count_dict[previous] = (success,total+1)

In [62]:
count_dict

{'SSS': (25, 58),
 'SSN': (36, 75),
 'SNS': (36, 75),
 'SNN': (39, 79),
 'NSS': (33, 75),
 'NSN': (39, 79),
 'NNS': (39, 78),
 'NNN': (39, 78)}

In [72]:
import pandas as pd
transition_matrix = pd.DataFrame(count_dict).T
transition_matrix.columns = ["Sunny", "Total"]
transition_matrix["Not Sunny"] = (transition_matrix["Total"] - transition_matrix["Sunny"])/transition_matrix["Total"]
transition_matrix["Sunny"] = transition_matrix["Sunny"]/transition_matrix["Total"]
transition_matrix= transition_matrix.drop(["Total"],axis=1)
transition_matrix

Sunny  Not Sunny
SSS  0.431034   0.568966
SSN  0.480000   0.520000
SNS  0.480000   0.520000
SNN  0.493671   0.506329
NSS  0.440000   0.560000
NSN  0.493671   0.506329
NNS  0.500000   0.500000
NNN  0.500000   0.500000

# ADD MORE DETAIL

#### Question 3

(15) Assume a Gaussian distribution for observations, Xi, i = 1, . . . , N with
unknown mean, M, and known variance 5. Suppose the prior for M is Gaussian
with variance 10. How large a random sample must be taken (i.e., what is the
minimum value for N) to specify an interval having unit length of 1 such that
the probability that M lies in this interval is 0.95?

#### Solution 3